server-common/src/main/java/org/apache/kafka/server/fault/ProcessTerminatingFaultHandler.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.server.fault;
+
+import java.util.Objects;
+import org.apache.kafka.common.utils.Exit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a fault handler which terminates the JVM process.
+ */
+final public class ProcessTerminatingFaultHandler implements FaultHandler {
+    private static final Logger log = LoggerFactory.getLogger(ProcessTerminatingFaultHandler.class);
+
+    private final Runnable action;
+    private final boolean shouldHalt;
+
+    private ProcessTerminatingFaultHandler(boolean shouldHalt, Runnable action) {
+        this.shouldHalt = shouldHalt;
+        this.action = action;
+    }
+
+    @Override
+    public RuntimeException handleFault(String failureMessage, Throwable cause) {
+        if (cause == null) {
+            log.error("Encountered fatal fault: {}", failureMessage);
+        } else {
+            log.error("Encountered fatal fault: {}", failureMessage, cause);
+        }
+
+        try {
+            action.run();
+        } catch (Throwable e) {
+            log.error("Failed to run terminating action.", e);
+        }
+
+        int statusCode = 1;
+        if (shouldHalt) {
+            Exit.halt(statusCode);
+        } else {
+            Exit.exit(statusCode);
+        }
+
+        return null;
+    }
+
+    public static final class Builder {
+        private boolean shouldHalt = false;
```

Comment:
I agree. I'll change the default to true. I want to keep the option to use exit in the future.

We can fix the KRaft IO thread if we don't wait on the `exit` call. We can do this if we call `exit` from a daemon thread and let the KRaft IO thread terminate immediately.

The KRaft controller thread is a little bit more complicated since fatal exception are handled in a few places in the code.